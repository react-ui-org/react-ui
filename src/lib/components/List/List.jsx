import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import styles from './List.scss';

export const List = ({
  align,
  autoWidth,
  children,
}) => {
  if (!children) {
    return null;
  }

  let alignClass;

  if (align === 'start') {
    alignClass = styles.alignStart;
  } else if (align === 'end') {
    alignClass = styles.alignEnd;
  }

  let autoWidthClass;

  if (autoWidth) {
    autoWidthClass = styles.isAutoWidth;
  }

  return (
    <div
      className={classNames(
        styles.root,
        autoWidthClass,
      )}
    >
      <ul
        className={classNames(
          styles.list,
          alignClass,
        )}
      >
        {children}
      </ul>
    </div>
  );
};

List.defaultProps = {
  align: 'start',
  autoWidth: false,
  children: null,
};

List.propTypes = {
  /**
   * Horizontal alignment of list items.
   */
  align: PropTypes.oneOf(['start', 'end']),
  /**
   * If `true`, the list items will take up only as much horizontal space as necessary.
   */
  autoWidth: PropTypes.bool,
  /**
   * Individual ListItems.
   */
  children: PropTypes.node,
};

export const ListWithContext = withProviderContext(List, 'List');

export default ListWithContext;
