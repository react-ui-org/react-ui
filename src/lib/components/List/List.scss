@use "../../styles/tools/reset";
@use "../../styles/tools/spacing";
@use "theme";

.root {
    @include spacing.bottom(layouts);
}

.list {
    @include reset.list();

    display: flex;
    flex-direction: column;
}

.item {
    width: 100%;
    margin-bottom: theme.$row-gap;

    &:last-child {
        margin-bottom: 0;
    }
}

.alignStart {
    align-items: flex-start;
    text-align: left;
}

.alignEnd {
    align-items: flex-end;
    text-align: right;
}

[dir="rtl"] .alignStart {
    text-align: right;
}

[dir="rtl"] .alignEnd {
    text-align: left;
}

.isAutoWidth .list {
    display: inline-flex;
}

.isAutoWidth .item {
    width: auto;
}

.item .root {
    margin-bottom: 0;
}
