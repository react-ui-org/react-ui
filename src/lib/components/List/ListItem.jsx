import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import styles from './List.scss';

export const ListItem = ({ children }) => {
  if (!children) {
    return null;
  }

  return (
    <li className={styles.item}>
      {children}
    </li>
  );
};

ListItem.defaultProps = {
  children: null,
};

ListItem.propTypes = {
  /**
   * Content of the list item.
   */
  children: PropTypes.node,
};

export const ListItemWithContext = withProviderContext(ListItem, 'ListItem');

export default ListItemWithContext;
