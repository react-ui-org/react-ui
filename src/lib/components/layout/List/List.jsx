import PropTypes from 'prop-types';
import React from 'react';
import styles from './List.scss';

export const List = ({
  align,
  autoWidth,
  children,
}) => {
  if (!children) {
    return null;
  }

  let alignClass = '';

  if (align === 'start') {
    alignClass = styles.alignStart;
  } else if (align === 'end') {
    alignClass = styles.alignEnd;
  }

  let autoWidthClass = '';

  if (autoWidth) {
    autoWidthClass = styles.isAutoWidth;
  }

  return (
    <div
      className={`
        ${styles.root}
        ${autoWidthClass}
      `.trim()}
    >
      <ul
        className={`
          ${styles.list}
          ${alignClass}
        `.trim()}
      >
        {children}
      </ul>
    </div>
  );
};

List.defaultProps = {
  align: 'start',
  autoWidth: false,
  children: null,
};

List.propTypes = {
  /**
   * Horizontal alignment of list items.
   */
  align: PropTypes.oneOf(['start', 'end']),
  /**
   * If `true`, the list items will take up only as much horizontal space as necessary.
   */
  autoWidth: PropTypes.bool,
  /**
   * Individual ListItems.
   */
  children: PropTypes.node,
};

export default List;
