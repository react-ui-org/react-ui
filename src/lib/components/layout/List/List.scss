@import '../../../styles/settings/layouts';
@import '../../../styles/tools/spacing';

.root {
  margin-bottom: $layout-common-bottom-spacing;
}

.list {
  display: flex;
  flex-direction: column;
  padding: 0;
  margin-bottom: 0;
  list-style: none;
}

.item {
  width: 100%;
  margin-bottom: spacing(2);

  &:last-child {
    margin-bottom: 0;
  }
}

.alignLeft {
  align-items: flex-start;
}

.alignRight {
  align-items: flex-end;
  text-align: right;
}

.isAutoWidth .list {
  display: inline-flex;
}

.isAutoWidth .item {
  width: auto;
}

.item .root {
  margin-bottom: 0;
}
