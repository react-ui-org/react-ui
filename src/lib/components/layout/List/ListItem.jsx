import PropTypes from 'prop-types';
import React from 'react';
import styles from './List.scss';

const ListItem = ({ children }) => {
  if (!children) {
    return null;
  }

  return (
    <li className={styles.item}>
      {children}
    </li>
  );
};

ListItem.defaultProps = {
  children: null,
};

ListItem.propTypes = {
  children: PropTypes.node,
};

export default ListItem;
