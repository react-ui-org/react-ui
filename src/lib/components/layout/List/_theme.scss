$row-gap: var(--rui-List__row-gap);
