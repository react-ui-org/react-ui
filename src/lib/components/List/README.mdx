---
name: List
menu: 'Layouts'
route: /components/list
---

# List

import {
  Playground,
  Props,
} from 'docz'
import { Placeholder } from '../../../docs/_components/Placeholder/Placeholder'
import { List } from './List'
import { ListItem } from './ListItem'

The List aligns content into an organized list.

## Basic Usage

To implement the List component, you need to import it first:

```js
import { List, ListItem } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <List>
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
  </List>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- **Wrap your items** into the [ListItem](#listitem) component. This ensures
  your content is properly spaced and aligned with other List elements. Do
  **not** try to put any custom HTML or React components directly into the
  List layout without wrapping it with the ListItem first.

- For forms, use rather the [FormLayout](/components/form-layout)
  component which is designed specifically for that purpose.

## List Alignment

You can simply set the list alignment by specifying the `align` option.

ðŸ‘‰ At the current stage of development, React UI is **RTL aware** so switching
to a fully RTL-compatible behavior in the future should be painless. That's why
the alignment values are called rather `start` than `left` and `end` instead of
`right`.

<Playground>
  <List align="end">
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
  </List>
</Playground>

## Auto Width

For cases when you prefer the list items to keep their original width and don't
want them to expand over the full width of the parent of the list, there is the
`autoWidth` option. It **shrinks the list according to its longest item.**
However, the root div always behaves as a full-width block element (unless you
put it inside a flex or grid layout).

<Playground>
  <List align="end" autoWidth>
    <ListItem>
      <Placeholder bordered>List item</Placeholder>
    </ListItem>
    <ListItem>
      <Placeholder bordered>List item that is longer</Placeholder>
    </ListItem>
    <ListItem>
      <Placeholder bordered>List item that is even a bit longer</Placeholder>
    </ListItem>
  </List>
</Playground>

## API

<Props table of={List} />

### ListItem API

A wrapper for individual list items.

<Props table of={ListItem} />

## Theming

| Custom Property                                      | Description                                                  |
|------------------------------------------------------|--------------------------------------------------------------|
| `--rui-List__row-gap`                                | Gap between list items                                       |
