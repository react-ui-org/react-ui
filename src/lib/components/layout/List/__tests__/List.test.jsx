import React from 'react';
import { shallow } from 'enzyme';
import { List } from '../List';
import { ListItem } from '../ListItem';

describe('rendering', () => {
  it('renders correctly with no children', () => {
    const tree = shallow((
      <List />
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a single child', () => {
    const tree = shallow((
      <List>
        <ListItem>content</ListItem>
      </List>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children', () => {
    const tree = shallow((
      <List>
        <ListItem>content 1</ListItem>
        <ListItem>content 2</ListItem>
        <ListItem>content 3</ListItem>
      </List>
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with multiple children and all props', () => {
    const tree = shallow((
      <List align="end" autoWidth>
        <ListItem>content 1</ListItem>
        <ListItem>content 2</ListItem>
        <ListItem>content 3</ListItem>
      </List>
    ));

    expect(tree).toMatchSnapshot();
  });
});
