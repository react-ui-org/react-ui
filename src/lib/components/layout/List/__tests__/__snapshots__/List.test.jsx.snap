// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with a single child 1`] = `
<div
  className="root"
>
  <ul
    className="list
          alignLeft"
  >
    <ListItem>
      content
    </ListItem>
  </ul>
</div>
`;

exports[`rendering renders correctly with multiple children 1`] = `
<div
  className="root"
>
  <ul
    className="list
          alignLeft"
  >
    <ListItem>
      content 1
    </ListItem>
    <ListItem>
      content 2
    </ListItem>
    <ListItem>
      content 3
    </ListItem>
  </ul>
</div>
`;

exports[`rendering renders correctly with multiple children and all props 1`] = `
<div
  className="root
        isAutoWidth"
>
  <ul
    className="list
          alignRight"
  >
    <ListItem>
      content 1
    </ListItem>
    <ListItem>
      content 2
    </ListItem>
    <ListItem>
      content 3
    </ListItem>
  </ul>
</div>
`;

exports[`rendering renders correctly with no children 1`] = `""`;
